### PR TITLE
Adds helper methods to grab a LedgerTx or verify a SignedTx.

### PR DESCRIPTION
**Pull request by joeldudleyr3**

### CODE REVIEW

File: core/src/main/kotlin/net/corda/core/transactions/SignedTransaction.kt [hunk 1]
```diff
@@ -146,7 +146,26 @@ data class SignedTransaction(val txBits: SerializedBytes<WireTransaction>,
      * @throws SignaturesMissingException if any signatures that should have been present are missing.
      */
     @Throws(AttachmentResolutionException::class, TransactionResolutionException::class, SignatureException::class)
-    fun toLedgerTransaction(services: ServiceHub) = verifySignatures().toLedgerTransaction(services)
+    fun toLedgerTransaction(services: ServiceHub, checkSufficientSignatures: Boolean = true): LedgerTransaction {
```
Comment (mikehearn) [inline core/src/main/kotlin/net/corda/core/transactions/SignedTransaction.kt]: Maybe use `@JvmOverloads` here to avoid an API break for java devs?

File: core/src/main/kotlin/net/corda/core/transactions/SignedTransaction.kt [hunk 2]
```diff
@@ -146,7 +146,26 @@ data class SignedTransaction(val txBits: SerializedBytes<WireTransaction>,
...
+     * [WireTransaction.toLedgerTransaction] with the passed in [ServiceHub] to resolve the dependencies and
+     * return an unverified LedgerTransaction, then verifies the LedgerTransaction.
+     *
+     * @throws AttachmentResolutionException if a required attachment was not found in storage.
+     * @throws TransactionResolutionException if an input points to a transaction not found in storage.
+     * @throws SignatureException if any signatures were invalid or unrecognised
+     * @throws SignaturesMissingException if any signatures that should have been present are missing.
+     */
+    @Throws(AttachmentResolutionException::class, TransactionResolutionException::class, SignatureException::class)
+    fun verify(services: ServiceHub, checkSufficientSignatures: Boolean = true) {
```
Comment (mikehearn) [inline core/src/main/kotlin/net/corda/core/transactions/SignedTransaction.kt]: Ditto

File: core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt [hunk 3]
```diff
@@ -138,6 +139,9 @@ open class TransactionBuilder(
     fun toWireTransaction() = WireTransaction(ArrayList(inputs), ArrayList(attachments),
             ArrayList(outputs), ArrayList(commands), notary, signers.toList(), type, timeWindow)
 
+    @Throws(AttachmentResolutionException::class, TransactionResolutionException::class)
```
Comment (mnesbit) [inline core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt]: Do you want to mirror the signed transaction and add a verify method?

**Comment by joeldudleyr3** [comment]
Thanks Matt, good idea, addressed.

**Comment by joeldudleyr3** [comment]
Added a `TransactionBuilder.verify` method. Added a `@Throws` for `TransactionVerificationException`.

Back to you.

File: core/src/main/kotlin/net/corda/core/transactions/SignedTransaction.kt [hunk 4]
```diff
@@ -136,17 +137,42 @@ data class SignedTransaction(val txBits: SerializedBytes<WireTransaction>,
...
+     * with the passed in [ServiceHub] to resolve the dependencies, returning an unverified
+     * LedgerTransaction.
      *
      * @throws AttachmentResolutionException if a required attachment was not found in storage.
      * @throws TransactionResolutionException if an input points to a transaction not found in storage.
      * @throws SignatureException if any signatures were invalid or unrecognised
      * @throws SignaturesMissingException if any signatures that should have been present are missing.
      */
-    @Throws(AttachmentResolutionException::class, TransactionResolutionException::class, SignatureException::class)
-    fun toLedgerTransaction(services: ServiceHub) = verifySignatures().toLedgerTransaction(services)
```
Comment (roger3cev) [inline core/src/main/kotlin/net/corda/core/transactions/SignedTransaction.kt]: could this be made private? Then we can change the `WireTransaction` to vend input `StateAndRef`s and attachments?

File: core/src/main/kotlin/net/corda/core/transactions/SignedTransaction.kt [hunk 5]
```diff
@@ -136,17 +137,42 @@ data class SignedTransaction(val txBits: SerializedBytes<WireTransaction>,
...
+    @JvmOverloads
+    @Throws(SignatureException::class, AttachmentResolutionException::class, TransactionResolutionException::class)
+    fun toLedgerTransaction(services: ServiceHub, checkSufficientSignatures: Boolean = true): LedgerTransaction {
```
Comment (roger3cev) [inline core/src/main/kotlin/net/corda/core/transactions/SignedTransaction.kt]: do we really need the `checkSufficientSignatures` flag. Surely we abstract this away and just check whatever signatures we have?

File: core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt [hunk 6]
```diff
@@ -149,6 +151,14 @@ open class TransactionBuilder(
         return SignedTransaction(wtx.serialize(), ArrayList(currentSigs))
     }
 
+    @Throws(AttachmentResolutionException::class, TransactionResolutionException::class)
+    fun toLedgerTransaction(services: ServiceHub) = toWireTransaction().toLedgerTransaction(services)
```
Comment (roger3cev) [inline core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt]: do we need this?

File: core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt [hunk 7]
```diff
@@ -197,10 +123,86 @@ open class TransactionBuilder(
...
+    fun attachments(): List<SecureHash> = ArrayList(attachments)
     fun outputStates(): List<TransactionState<*>> = ArrayList(outputs)
     fun commands(): List<Command> = ArrayList(commands)
-    fun attachments(): List<SecureHash> = ArrayList(attachments)
+
+    /** The signatures that have been collected so far - might be incomplete! */
+    @VisibleForTesting
+    protected val currentSigs = arrayListOf<DigitalSignature.WithKey>()
+
+    @VisibleForTesting
```
Comment (roger3cev) [inline core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt]: this was deprecated before - we should leave it as deprecated

File: core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt [hunk 8]
```diff
@@ -197,10 +123,86 @@ open class TransactionBuilder(
...
+     */
+    @VisibleForTesting
+    fun checkSignature(sig: DigitalSignature.WithKey) {
+        require(commands.any { it.signers.any { sig.by in it.keys } }) { "Signature key doesn't match any command" }
+        sig.verify(toWireTransaction().id)
+    }
+
+    /** Adds the signature directly to the transaction, without checking it for validity. */
+    @VisibleForTesting
+    fun addSignatureUnchecked(sig: DigitalSignature.WithKey): TransactionBuilder {
```
Comment (roger3cev) [inline core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt]: `signInitialTransaction()` uses this

Comment (joeldudleyr3) [inline core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt]: I've rewritten `signInitialTransaction` accordingly.

File: core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt (hunk 7 above)
Comment (joeldudleyr3) [inline core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt]: OK - everything is marked `@Deprecated` now instead.

**Comment by mikehearn** [comment]
This is currently approved but failing TeamCity. Joel, are you adding more code to the PR after it's been reviewed?

**Comment by joeldudleyr3** [comment]
Hey Mike - yep. This is the one we spoke about refactoring the tests for to remove the uses of deprecated methods. I'll do a re-review afterwards.

File: core/src/main/kotlin/net/corda/core/node/ServiceHub.kt [hunk 9]
```diff
@@ -2,12 +2,15 @@ package net.corda.core.node
 
 import net.corda.core.contracts.*
 import net.corda.core.crypto.DigitalSignature
+import net.corda.core.crypto.isFulfilledBy
```
Comment (mnesbit) [inline core/src/main/kotlin/net/corda/core/node/ServiceHub.kt]: Is this import used?

File: core/src/main/kotlin/net/corda/core/transactions/SignedTransaction.kt [hunk 10]
```diff
@@ -136,17 +138,42 @@ data class SignedTransaction(val txBits: SerializedBytes<WireTransaction>,
     operator fun plus(sigList: Collection<DigitalSignature.WithKey>) = withAdditionalSignatures(sigList)
 
     /**
-     * Calls [verifySignatures] to check all required signatures are present, and then calls
-     * [WireTransaction.toLedgerTransaction] with the passed in [ServiceHub] to resolve the dependencies,
-     * returning an unverified LedgerTransaction.
+     * Checks the transaction's signatures are valid, optionally calls [verifySignatures] to check
```
Comment (mnesbit) [inline core/src/main/kotlin/net/corda/core/transactions/SignedTransaction.kt]: Perhaps make it a bit clearer why you would use this.

File: finance/src/test/kotlin/net/corda/contracts/asset/CashTests.kt [hunk 11]
```diff
@@ -43,8 +43,8 @@ class CashTests {
             amount = Amount(amount.quantity, token = amount.token.copy(amount.token.issuer.copy(reference = OpaqueBytes.of(ref))))
     )
 
-    lateinit var services: MockServices
-    val vault: VaultService get() = services.vaultService
+    lateinit var miniCorpServices: MockServices
```
Comment (mnesbit) [inline finance/src/test/kotlin/net/corda/contracts/asset/CashTests.kt]: Can we have a TODO somewhere for us to re-evaluate/remove the minicorp/megacorp stuff, because it is unclear how that would map to the current key manager and identity services.

File: node/src/test/kotlin/net/corda/node/services/vault/VaultQueryTests.kt [hunk 12]
```diff
@@ -1007,7 +1007,7 @@ class VaultQueryTests {
 
             // DOCSTART VaultDeprecatedQueryExample3
```
Comment (mnesbit) [inline node/src/test/kotlin/net/corda/node/services/vault/VaultQueryTests.kt]: Are these changes just whitespace only? If so please revert

File: core/src/main/kotlin/net/corda/core/node/ServiceHub.kt (hunk 9 above)
Comment (joeldudleyr3) [inline core/src/main/kotlin/net/corda/core/node/ServiceHub.kt]: addressed

File: core/src/main/kotlin/net/corda/core/transactions/SignedTransaction.kt (hunk 10 above)
Comment (joeldudleyr3) [inline core/src/main/kotlin/net/corda/core/transactions/SignedTransaction.kt]: addressed

File: finance/src/test/kotlin/net/corda/contracts/asset/CashTests.kt (hunk 11 above)
Comment (joeldudleyr3) [inline finance/src/test/kotlin/net/corda/contracts/asset/CashTests.kt]: addressed

File: node/src/test/kotlin/net/corda/node/services/vault/VaultQueryTests.kt (hunk 12 above)
Comment (joeldudleyr3) [inline node/src/test/kotlin/net/corda/node/services/vault/VaultQueryTests.kt]: addressed

File: core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt [hunk 13]
```diff
@@ -187,10 +120,84 @@ open class TransactionBuilder(
...
+     * The signature can be obtained using [NotaryFlow].
+     *
+     * The window of time in which the final time-window may lie is defined as [time] +/- [timeTolerance].
+     * If you want a non-symmetrical time window you must add the command via [addCommand] yourself. The tolerance
+     * should be chosen such that your code can finish building the transaction and sending it to the TSA within that
+     * window of time, taking into account factors such as network latency. Transactions being built by a group of
+     * collaborating parties may therefore require a higher time tolerance than a transaction being built by a single
+     * node.
+     */
+    fun addTimeWindow(time: Instant, timeTolerance: Duration) = addTimeWindow(TimeWindow.withTolerance(time, timeTolerance))
```
Comment (mnesbit) [inline core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt]: The design doc suggested setTimeWindow. It might make more sense

Comment (joeldudleyr3) [inline core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt]: This clashes with the auto-generated Kotlin set for `this.timeWindow`, unfortunately.

Comment (mikehearn) [inline core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt]: Why? Don't they have different signatures? You should be able to call it setTimeWindow as far as I can see ...

Comment (joeldudleyr3) [inline core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt]: It definitely seems to be clashing. If I take off the `protected` keyword and access it from Java, a `setTimeWindow` method has been generated.

Comment (mikehearn) [inline core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt]: How are you deciding it is clashing? Do you get an error from the compiler?

Comment (joeldudleyr3) [inline core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt]: Yes - it underlines the method and the constructor param, and says they have the same declaration.

Comment (mikehearn) [inline core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt]: Ah, I see what's going on here.

This review thread is attached to the following line of code:

`fun addTimeWindow(time: Instant, timeTolerance: Duration) = addTimeWindow(TimeWindow.withTolerance(time, timeTolerance))`

That function has two parameters and will not clash if renamed.

A little bit below is this method:

`fun addTimeWindow(timeWindow: TimeWindow)`

which will clash because it indeed has the same signature. However, note that we can rewrite the code like this:

```
open class TransactionBuilder(...., timeWindow: TimeWindow? = null) {
    protected var timeWindow = timeWindow
        set {
            check(notary != null) { "Only notarised transactions can have a time-window" }
            signers.add(notary!!.owningKey)
            this.timeWindow = timeWindow
        }
}
```

Comment (joeldudleyr3) [inline core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt]: Makes sense, will make a separate PR.
